client/webserver/site/src/js/markets.js

```diff
@@ -1525,53 +1557,103 @@ export default class MarketsPage extends BasePage {
-    tr.order = order
-    const rate = order.rate
+    const { rate, sell } = orderBin[0]
+    const isEpoch = !!orderBin[0].epoch
+    tr.orderBin = orderBin
     bind(tr, 'click', () => {
       this.reportClick(rate)
     })
     let qtyTD
+    const updateQtyTD = () => {
+      const mainDiv = qtyTD.children[0]
+      const [qtyDiv, ordersDiv] = mainDiv.children
+      const qty = tr.orderBin.reduce((total, curr) => total + curr.qty, 0)
+      qtyDiv.innerText = qty.toFixed(8)
+      ordersDiv.innerText = tr.orderBin.length
+      ordersDiv.title = `quantity comprised of ${tr.orderBin.length} ` +
+        (tr.orderBin.length > 1 ? 'orders' : 'order')
+    }
     tr.querySelectorAll('td').forEach(td => {
       switch (td.dataset.type) {
         case 'qty':
           qtyTD = td
-          td.innerText = order.qty.toFixed(8)
+          updateQtyTD()
           break
         case 'rate':
-          if (order.rate === 0) {
+          if (rate === 0) {
             td.innerText = 'market'
           } else {
-            td.innerText = order.rate.toFixed(8)
+            td.innerText = rate.toFixed(8)
             td.classList.add(cssClass)
             // Draw a line on the chart on hover.
             Doc.bind(tr, 'mouseenter', e => {
               const chart = this.chart
               this.depthLines.hover = [{
-                rate: order.rate,
-                color: order.sell ? chart.theme.sellLine : chart.theme.buyLine
+                rate: rate,
+                color: sell ? chart.theme.sellLine : chart.theme.buyLine
               }]
               this.drawChartLines()
             })
           }
           break
         case 'epoch':
-          if (order.epoch) td.appendChild(check.cloneNode())
+          if (isEpoch) td.appendChild(check.cloneNode())
           break
       }
     })
-    tr.addQty = (qty) => {
-      tr.qty += qty
-      qtyTD.innerText = tr.qty.toFixed(8)
+    tr.insertOrder = (order) => {
+      tr.orderBin.push(order)
+      updateQtyTD()
+    }
+    tr.updateOrderQty = (token, qty) => {
+      for (let i = 0; i < tr.orderBin.length; i++) {
+        if (tr.orderBin[i].token === token) {
+          tr.orderBin[i].qty = qty
+          updateQtyTD()
+          return true
+        }
+      }
+      return false
+    }
+    tr.removeOrder = (token) => {
+      const index = tr.orderBin.findIndex(order => order.token === token)
+      if (index < 0) return false
+      tr.orderBin.splice(index, 1)
+      if (!tr.orderBin.length) tr.remove()
+      else updateQtyTD()
+      return true
+    }
+    tr.removeEpochOrders = (newEpoch) => {
+      tr.orderBin = tr.orderBin.filter((order) => {
+        return !(order.epoch && order.epoch !== newEpoch)
+      })
+      if (!tr.orderBin.length) tr.remove()
+      else updateQtyTD()
+    }
+    tr.getRate = () => {
+      return tr.orderBin[0].rate
+    }
+    tr.getEpoch = () => {
+      return tr.orderBin[0].epoch
+    }
+    tr.compare = (order) => {
```

Comment:
This is good stuff, and it makes the semantics of updating these rows really clean, but once we're adding seven new functions as custom properties to an `td` element, that's a big clue to me that we should just write a custom class to manage the element and make these class methods.

Reply:
I've added a manager class that has a reference to the `td` element the `td` element has a reference to the manager class. It's definitely a lot cleaner.